src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String EXCEPTION_MESSAGE = "Provided index is out of valid range";
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndex(index);
+        Node<T> newNode = new Node<T>(null, value, null);
+        if (index == 0 && size == 0) {
+            head = newNode;
+            tail = newNode;
+        } else if (index == 0) {
+            head.prev = newNode;
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> sameIndexNode = getNode(index);
+            newNode.prev = sameIndexNode.prev;
+            sameIndexNode.prev.next = newNode;
+            sameIndexNode.prev = newNode;
+            newNode.next = sameIndexNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
```

Comment:
the index need to be checked in the get() method

Reply:
the getNode() method has index validation inside it

Reply:
I saw it, but it would be more logical to put it in get. I guess